mlflow/pyfunc/__init__.py

```diff
@@ -1648,21 +1662,175 @@ def _convert_struct_values(
+    prebuilt_env_sha, prebuilt_runtime_version, prebuilt_platform_machine = archive_name.split("-")[
+        -3:
+    ]
+
+    python_env = _get_python_env(Path(local_model_path))
+    env_sha = _get_virtualenv_name(python_env, local_model_path).split("-")[-1]
+    dbconnect_cache = get_dbconnect_client_cache(spark)
+    runtime_version = dbconnect_cache.udf_sandbox_image_version
+    platform_machine = dbconnect_cache.udf_sandbox_platform_machine
+
+    if prebuilt_env_sha != env_sha:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' does not match the model required environment."
+        )
+    if prebuilt_runtime_version != runtime_version:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' runtime version '{prebuilt_runtime_version}' "
+            f"does not match UDF sandbox runtime version {runtime_version}."
+        )
+    if prebuilt_platform_machine != platform_machine:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' platform machine '{prebuilt_platform_machine}' "
+            f"does not match UDF sandbox platform machine {platform_machine}."
+        )
+
+
+def _prebuild_env_internal(local_model_path, archive_name, save_path):
+    env_root_dir = os.path.join(_PREBUILD_ENV_ROOT_LOCATION, archive_name)
+    archive_path = os.path.join(save_path, archive_name + ".tar.gz")
+    if os.path.exists(env_root_dir):
+        shutil.rmtree(env_root_dir)
+    if os.path.exists(archive_path):
+        os.remove(archive_path)
+
     try:
-        from pyspark.sql.utils import is_remote
-    except ImportError:
-        return False
-    return is_remote()
+        pyfunc_backend = get_flavor_backend(
+            local_model_path,
+            env_manager="virtualenv",
+            install_mlflow=False,
+            create_env_root_dir=False,
+            env_root_dir=env_root_dir,
+        )
+
+        pyfunc_backend.prepare_env(model_uri=local_model_path, capture_output=False)
+        # exclude pip cache from the archive file.
+        shutil.rmtree(os.path.join(env_root_dir, "pip_cache_pkgs"))
+
+        return archive_directory(env_root_dir, archive_path)
+    finally:
+        shutil.rmtree(env_root_dir, ignore_errors=True)
+
+
+def build_model_env(model_uri, save_path):
+    """
+    Prebuild model python environment and generate an archive file saved to provided
+    `save_path`.
+
+    Typical usages:
+     - Prebuild the model environment in Databricks Runtime, and then download the prebuilt
+       python environment archive file, then the prebuilt env file can be used
+       in `mlflow.pyfunc.spark_udf` in any remote Databricks connect client side.
+     - When running `mlflow.pyfunc.spark_udf` in Databricks runtime with `env_manager='virtualenv',
+       setting `prebuilt_env_path` can speed up `mlflow.pyfunc.spark_udf` execution because
+       it can skip rebuilding the python environment.
```

Comment:
Does this mean `build_model_env + spark_udf` is faster than `spark_udf`?

Reply:
spark_udf using prebuilt env is faster. (excluding the time of building env)

Reply:
what if the time of building env is included?

Reply:
then it will be slightly slower (because packaging the env and upload env to NFS consumes additional times)

but note "building env" only happens once and can be used across different sessions, so it deserves doing this.

Reply:
it's faster or not?

Reply:
If not, we should remove it.